itou/communications/cache.py

```diff
@@ -5,23 +5,30 @@
 from itou.communications.models import AnnouncementCampaign
 
 
-CACHE_ACTIVE_ANNOUNCEMENT_CAMPAIGN_KEY = "active-announcement-campaign"
+CACHE_ACTIVE_ANNOUNCEMENTS_KEY = "active-announcement-campaign"
 
 
 def update_active_announcement_cache():
     today = date.today()
     last_edition_boundary = today.replace(day=1) - timedelta(days=1)
 
     campaign = (
-        AnnouncementCampaign.objects.filter(start_date__lte=today, start_date__gt=last_edition_boundary)
-        .prefetch_related("items")
-        .first()
+        AnnouncementCampaign.objects.filter(
+            start_date__lte=today, start_date__gt=last_edition_boundary
+        ).prefetch_related("items")
     )
 
-    if campaign is None:
-        cache.set(CACHE_ACTIVE_ANNOUNCEMENT_CAMPAIGN_KEY, {"value": None}, None)
-        return None
+    def get_cache_expiration():
+        if not len(campaign):
+            return None
+        return (datetime.combine(campaign[0].end_date, datetime.min.time()) - datetime.now()).total_seconds()
 
-    seconds_until_end = (datetime.combine(campaign.end_date, datetime.min.time()) - datetime.now()).total_seconds()
-    cache.set(CACHE_ACTIVE_ANNOUNCEMENT_CAMPAIGN_KEY, {"value": campaign}, seconds_until_end)
+    cache.set(CACHE_ACTIVE_ANNOUNCEMENTS_KEY, campaign, get_cache_expiration())
     return campaign
+
+
+def get_cached_active_announcement():
+    campaign = cache.get(CACHE_ACTIVE_ANNOUNCEMENTS_KEY)
+    if campaign is None:
+        campaign = update_active_announcement_cache()
+    return campaign.first()
```

Comment:
J'ai l'impression que la requête est faite ici du coup, if vaut mieux le faire dans `update_active_announcement_cache()`

Reply:
Je n'ai pas compris, quelle requête ? Le `if campaign is None` ici est parce que si le cache est vidé il faut le mettre à jour. Le fonction est un wrapper pour `cache.get`

Reply:
`campaign.first()` fait une requête (récupérer le 1er élément).
Il faut que tu stocke `campaign.first()` dans le cache pour que tu récupères directement ici l'objet

Reply:
J'ai un test normalement qui [confirme qu'il y a 0 requêtes ici](https://github.com/gip-inclusion/les-emplois/blob/340871d6b771a91eb05797175a0910e2fdee4601/tests/communications/test_cache.py#L14), peut-être parce que le `QuerySet.count() == 1`?

Originalement on stockait un dictionnaire dans le cache (sous la forme `value: campaign.first()`), mais on a utilisé le QuerySet en tant qu'alternative (`AnnouncementCampaign.none() != None`)

Reply:
Tiens, c'est étrange :thinking: 

En tout cas, j'ai l'impression que l'on peut avoir un problème : si fin aout il n'y a pas de campagne en cours, alors on va mettre un queryset "vide" dans le cache sans date de fin, et comme`campaign` ne vaudra pas `None`, `get_cached_active_announcement()` ne rechargera pas de nouvelle données le 1er septembre.
Garder le queryset "vide" est bien car cela permet de ne pas appeler `update_active_announcement_cache()` à chaque appel de `get_cached_active_announcement()`, par contre il faudrait peut être mettre un nombre de secondes jusqu'à la fin du mois en cours quand il n'y a pas de campagnes en cours (-> ligne 18).

Et tu peux renommer les `campaign` en `campaign_qs` pour qu'il soit clair que c'est un queryset.

Reply:
pour mon scénario en effet ça fonctionne bien, tu as raison

Reply:
Cela me semble étrange de stocker un queryset plutôt qu'un objet dans le cache.
Pour distinguer entre `None` indiquant pas de campagne et `None` indiquant un cache vide il suffit d'utiliser une autre valeur sentinel, cf https://docs.djangoproject.com/en/5.0/topics/cache/#django.core.cache.cache.get

Reply:
Je suis content de changer la solution d'utiliser un autre valeur sentinel